src/shared/ui/picker.ts

```diff
@@ -145,3 +149,201 @@ export function verifySinglePickerOutput<T extends vscode.QuickPickItem>(choices
+                    default:
+                        if (this.pickerOptions.onDidTriggerButton) {
+                            this.pickerOptions.onDidTriggerButton(button, resolve, reject)
+                        }
+                }
+            },
+        })
+    }
+
+    private async loadItems(): Promise<void> {
+        // unpause the loader (if it was paused previously by a selection)
+        this.isPaused = false
+
+        // use a while loop so we have greater control over the iterator.
+        // breaking out of a for..of loop for an iterator will automatically set the iterator to `done`
+        // manual iteration means that we can use the same iterator no matter how many times we call loadItems()
+        while (!this.isDone && !this.isPaused) {
+            this.quickPick.busy = true
+            try {
+                const item = await this.iterator.next()
+                if (!item) {
+                    this.isDone = true
+                    break
+                }
+                const nextItems = this.awsCallLogic.awsCallResponseToQuickPickItemFn(item.value)
+                this.items =
+                    this.pickerOptions.manualPaginationType === 'replace' ? nextItems : this.items.concat(nextItems)
+                // TODO: Is there a way to append to this ReadOnlyArray so it doesn't constantly pop focus back to the top?
+                this.quickPick.items = this.items
+                // if this is manually paginated, pause the load after one pull
+                if (this.pickerOptions.manualPaginationType) {
+                    this.isPaused = true
+                }
+                this.isDone = item.done ?? false
+            } catch (e) {
+                // append error node
+                // clicking error node should either go backwards (return undefined) or refresh
+                // maybe have one of each?
+                // give quickpick an error message
+                // we should not blow away the existing items
+                const err = e as Error
+                this.items.push({
+                    ...this.errorItem,
+                    detail: err.message,
+                })
+                this.quickPick.items = this.items
+            }
+        }
+
+        if (this.isDone && this.items.length === 0) {
+            this.items.push(this.noItemsItem)
+            this.quickPick.items = this.items
+        }
+        this.quickPick.busy = false
+    }
+
+    private refreshQuickPick(): void {
+        this.items = []
+        this.isDone = false
+        this.quickPick.items = this.items
+        this.iterator = this.createNewIterator()
+    }
+
+    /**
+     * Generates a new iterator. Used at construction and during a "refresh" scenario.
+     */
+    private createNewIterator(): AsyncIterator<TResponse> {
+        return getPaginatedAwsCallIter(this.awsCallLogic.iteratorParams)
```

Comment:
Here you are just calling functions on behalf of input provided by the caller. Look at what's going on with the input provided by the caller. What you actually need from the caller is a factory that creates iterators that return arrays of quick pick items.

I would suggest you to create an interface for this class like:

```typescript
interface PickerContent {
   items: vscode.QuickPickItem[]
}
```

Then you can simply make the input in the constructor be a function that returns an `AsyncIterator<PickerContent>`

* This class no longer knows *how* to create the iterator
* It no longer knows *how* the iterator produces its output
* The caller of this class no longer necessary needs to provide an AWS service call as the input. Anything that produces `PickerContent` will do

This decouples this class from the logic to produce items and the types of items it can produce which will make it more generic and easier to test.